scipy/stats/_hypotests.py

```diff
@@ -1393,3 +1396,674 @@ def cramervonmises_2samp(x, y, method='auto'):
+    considered. For statistics that do not depend on the order of the data
+    within samples, this dramatically reduces computational cost without
+    affecting the shape of the null distribution (because the frequency/count
+    of each value is affected by the same factor).
+
+    ``permutation_type='both'`` does not support one-sample statistics, but it
+    can be applied to statistics with more than two samples. In this case, if
+    ``n`` is an array of the number of observations within each sample, the
+    number of distinct partitions is::
+
+        np.product([binom(sum(n[i:]), sum(n[i+1:])) for i in range(len(n)-1)])
+
+    **Paired statistics, permute pairings** (``permutation_type='pairings'``):
+
+    The null hypothesis associated with this permutation type is that
+    observations within each sample are drawn from the same underlying
+    distribution and that pairings with elements of other samples are
+    assigned at random.
+
+    Suppose ``data`` contains only one sample; e.g. ``a, = data``, and we
+    wish to consider all possible pairings of elements of ``a`` with elements
+    of a second sample, ``b``. Let ``n`` be the number of observations in
+    ``a``, which must also equal the number of observations in ``b``.
+
+    When ``1 < permutations < factorial(n)``, the elements of ``a`` are
+    randomly permuted. The user-supplied statistic accepts one data argument,
+    say ``a_perm``, and calculates the statistic considering ``a_perm`` and
+    ``b``. This process is performed repeatedly, `permutation` times,
+    generating a distribution of the statistic under the null hypothesis.
+    The statistic of the original data is compared to this distribution to
+    determine the p-value.
+
+    When ``permutations >= factorial(n)``, an exact test is performed:
+    ``a`` is permuted in each distinct way exactly once. Therefore, the
+    `statistic` is computed for each unique pairing of samples between ``a``
+    and ``b`` exactly once.
+
+    ``permutation_type='pairings'`` supports ``data`` containing any number
+    of samples, each of which must contain the same number of observations.
+    All samples provided in ``data`` are permuted *independently*. Therefore,
+    if ``m`` is the number of samples and ``n`` is the number of observations
+    within each sample, then the number of permutations in an exact test is::
+
+        factorial(n)**m
+
+    Note that if a two-sample statistic, for example, does not inherently
+    depend on the order in which observations are provided - only on the
+    *pairings* of observations - then only one of the two samples should be
+    provided in ``data``. This dramatically reduces computational cost without
+    affecting the shape of the null distribution (because the frequency/count
+    of each value is affected by the same factor).
+
+    **Paired statistics, permute samples** (``permutation_type='samples'``):
```

Comment:
I did not understand the different null hypothesis if the data is paired. Maybe an example would help. Disclaimer: I am not very familiar with paired samples :)

Reply:
Suppose we have samples `a` and `b`.
```
a = [a1, a2, a3]
b = [b1, b2, b3]
```

With `permutation_type = 'samples'`, we get permutation samples `x` and `y` by permuting observations between samples while preserving the order within samples (and thus pairings), e.g.
```
x = [b1, a2, b3]
y = [a1, b2, a3]
```
This is what is going on with tests like the (paired) Wilcoxon signed-rank test and paired t-test.

With `permutation_type = 'pairings'`, we permute the order (pairings) but keep observations within the same sample, e.g.
```
x = [a1, a2, a3]
y = [b3, b1, b2]  # y has been permuted, so the elements are paired with those of `x` differently
```
This is what is going on in correlation tests.

With `permutation_type='both'`, both types of permutations are performed.
```
x = [b3, a2, b2]
y = [a1, b1, a3]
```
This is what is going on in independent sample tests like Mann-Whitney U and the independent sample t-test.

Should I include these examples in the notes?

Reply:
yes, that is helpful!